Feat/radiobutton: RadioButton 컴포넌트 구현

prayinforrain opened the pull request:
## 🤠 개요

<!--

- 이슈번호
- 한줄 설명

-->
- Closes #42
- 라디오 버튼을 구현했어요.

## 💫 설명

<!--

- 현재 Pr 설명

-->
기본적으로 `input[type="radio"]`에 필요한 속성 외에 아래와 같은 속성이 있어요.
- customButton: 라디오 버튼의 생김새를 커스텀 컴포넌트로 사용할 수 있어요. 아래 속성들은 기본 button에만 영향을 줘요. (outerSize 제외)
- color: 라디오 버튼의 색상을 지정해요.
- size: 라디오 버튼의 크기를 지정해요.
- outerSize: 클릭 가능한 영역이 실제 버튼의 크기보다 넓어야 하는 상황이 있을 거라고 생각했어요. 그래서 클릭 영역의 크기를 직접 지정할 수 있게 했어요.

원래는 `children`으로 `label`같은 것들을 넣을 수 있게 했었는데, 모양새가 너무 이상해서 지금의 형태로 바뀌었어요. story 파일에 구현되어 있는 것처럼 하나의 `label` 태그 안에 넣으면 원하는 모양으로 커스텀하기 쉬우니 그렇게 사용하도록 하는 것이 맞는 것 같아요.

그리고 스토리북에서 `Controls`를 이쁘게 정리하고 싶어서 이것저것 옵션을 적용해서 코드가 조금 길어졌어요..  

근데 RadioButton보다 Radio라는 이름이 더 나을까요?


## 📷 스크린샷 (Optional)
![image](https://user-images.githubusercontent.com/46566891/222966892-b52461cd-d0e2-4a6d-8142-7b7c0ca6ff4f.png)  
![image](https://user-images.githubusercontent.com/46566891/222967336-f63870df-8760-4fc4-8432-5db8ce27f06b.png)  
![image](https://user-images.githubusercontent.com/46566891/222967316-d1ac17b5-203a-444c-890b-dad3f7df2878.png)

prayinforrain committed:
feat: RadioButton 컴포넌트 구현

prayinforrain committed:
fix: story control에 outerSize가 누락됨

prayinforrain committed:
fix: CustomSize 스토리에서 outerSize가 잘못 설정됨

prayinforrain committed:
fix: flag -> button 이름 수정

leesunmin1231 commented on src/components/RadioButton/index.tsx:
```diff
@@ -0,0 +1,97 @@
+  }
+`;
+
+const Button = styled.svg`
```
우와 svg태그가 있네요? 대박 신기해요!!!!!

prayinforrain replied:
실제 svg 파일들도 vs코드로 열어보면 svg 태그로 되어있을거에요! 이거 복붙하면 svg 파일 없이 인라인으로 삽입할 수도 있어요  

원래 테두리div 안에 색칠된 div 넣으려했는데 정확히 가운데에 오질 않아서(이유불명) 화나서 svg로 했어요..

leesunmin1231 commented on src/components/RadioButton/RadioButton.stories.tsx:
```diff
@@ -0,0 +1,150 @@
+  align-items: center;
+`;
+
+const DUMMY_CUSTOM_FLAG = styled.div`
+  border: 1px solid black;
+  border-radius: 0.2rem;
+  background-color: white;
+  width: 1rem;
+  height: 1rem;
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  padding-left: 0.4rem;
+  padding-bottom: 0.4rem;
+
+  & > div {
+    display: none;
+  }
+
+  input[type='radio']:checked + & > div {
+    display: inline;
+  }
+
+  input[type='radio']:disabled + & {
+    filter: grayscale(100%);
+  }
+`;
```
헛 커스텀이 이정도까지 가능하군요..?
다른 컴포넌트도 좀 더 세부적인 커스텀까지 가능하게 만들어야할지 고민이 되네요..

prayinforrain replied:
아래 사진같은 라디오 버튼이 필요할까 해서 뚫어놨는데 의미가 있는지 모르겠어요.  
![image](https://user-images.githubusercontent.com/46566891/223013119-c979d2fd-22ed-4255-9d54-95a5e019a3ef.png)

prayinforrain committed:
Merge branch 'main' into feat/radiobutton

prayinforrain committed:
feat: input element에 id 전달

se030 commented on src/components/RadioButton/index.tsx:
```diff
@@ -0,0 +1,97 @@
+import { theme } from '@components/@common/CdsProvider/theme';
```
이거 useTheme 훅이나 색상만 사용하는거면 constants에 COLOR로 써주세요!

prayinforrain committed:
Merge branch 'main' into feat/radiobutton

se030 commented on src/components/RadioButton/RadioButton.stories.tsx:
```diff
@@ -0,0 +1,150 @@
+  },
+} as ComponentMeta<typeof RadioButton>;
+
+const DUMMY_LABEL = styled.label`
```
이런거 네이밍이 상수인데 컴포넌트라서 좀 헷갈려요 ..!😂 PascalCase로 하면 보기 더 편할 것 같아요!

iyu88 replied:
아 이거... 여러 다른 스토리에서 이런 컨벤션 사용한 것 같아요.
다른 스토리들도 다 한번씩 수정해야할 것 같아요

se030 commented on src/components/RadioButton/RadioButton.stories.tsx:
```diff
@@ -0,0 +1,150 @@
+    >
+      팀장뽑기
+    </legend>
+    <DUMMY_LABEL>
```
```html
<div>
  <input id="id" />
  <label for="id" />
</div>
```

span 대신 이렇게 사용하는게 더 좋지 않을까요?
그리고 라디오 버튼에는 보통 라벨 붙여서 많이 쓰니까 아예 컴포넌트 안으로 넣어버리면 어떤가요? props로 옵셔널 라벨 받아서요 ...!

prayinforrain replied:
지금 RadioButton 컴포넌트는 체크박스를 연출하기 위한 div를 클릭할 수 있도록 아래와 같은 구조로 되어 있어요.
```
<label>
  <input type="radio" /> // display:none;
  <svg/> // 체크박스
</label>
```

RadioButton 컴포넌트의 props로 id를 받아 input 태그로 내려주고 있어서 `<label for="">`로 라벨을 붙이는 것도 가능하고, 스토리 예제처럼 하나의 label 태그 안에 넣어서 사용도 가능하기 때문에 굳이 라벨을 받아서 구현할 필요가 없다고 생각했어요.  
그리고 라벨의 경우에 단순히 string으로 받을 게 아니라 문자열, 아이콘이나 그림 등등이 들어갈 수 있어서 컴포넌트로 받아야 하는데 사용 케이스를 생각해보면 라디오버튼 자체만 구현해 두고 라벨 부분은 따로 구현하는 것이 나을 것이라고 생각했습니다.(라벨과 버튼의 위치도 조절할 수 있게 해야하고.. 어쩌고저쩌고..)

se030 commented on src/components/RadioButton/RadioButton.stories.tsx:
```diff
@@ -0,0 +1,150 @@
+  padding-left: 0.4rem;
+  padding-bottom: 0.4rem;
+
+  & > div {
```
그리고 커스텀 되는건 너무 좋은데 커스텀 스타일을 작성하기 위해서 내부 동작을 전부 뜯어보는건 번거로운 일 같아서요.
이런 컴포넌트 자체의 동작 같은건 내부에 * selector로 정의하거나, 커스텀을 더 지원하려면 customDefault / Checked / Disabled 같이 props로 받으면 더 좋을 것 같아요!

이런 애들?
```css
  & > div {
    display: none;
  }
  input[type='radio']:checked + & > div {
    display: inline;
  }
  input[type='radio']:disabled + & {
    filter: grayscale(100%);
  }
```

prayinforrain replied:
저도 커스텀 버튼을 구현해보면서 비슷한 문제를 느꼈어요. 한 번 방법을 찾아 볼게요..

prayinforrain committed:
fix: color useTheme로 변경

prayinforrain committed:
Merge branch 'feat/radiobutton' of https://github.com/c-h-w-h/cds into feat/radiobutton

prayinforrain committed:
style: 상수가 아닌 템플릿용 컴포넌트 네이밍 변경

prayinforrain committed:
Merge branch 'main' into feat/radiobutton

prayinforrain commented:
draft라는 개념이 있는걸 최근에 알았어요.  
원래는 아직 작업중인 PR이라는 의미로 리뷰를 막는 거긴 한데.. 최근에 본 여러 레퍼런스 + 오늘 회의하면서 변경해야 할 부분이 꽤 많은 것 같아서 일단 draft로 돌려놓고 나중에 취소하고 새 PR을 올리든 이 브랜치에서 리팩토링하든 할게요.

dohun31 committed:
Merge branch 'main' into feat/radiobutton

prayinforrain committed:
Merge branch 'main' into feat/radiobutton

prayinforrain committed:
refactor: RadioButton wrapper div로 변경

prayinforrain committed:
refactor: 라디오버튼에 z-index 추가

커스텀 버튼 사용시의 작동 오류를 방지하기 위해 z-index를 명시했습니다.

prayinforrain committed:
docs: RadioButton docs 작성

prayinforrain committed:
fix: code 탭 가독성 개선

prayinforrain committed:
chore: stories 파일 정리

prayinforrain committed:
feat: 접근성 속성 추가

- 실제 기능이 없는 button을 disable하였습니다.
- 그 외의 접근성 기능은 input 태그가 대신하므로 따로 지정하지 않았습니다.
- 추가로 inputRef에 대한 argTypes를 지정하였습니다.

prayinforrain commented:
이제서야 조금 준비가 되어가고 있어요. 사실 코드상 변화는 별로 없어요. 가장 큰 변화는 두 가지 정도에요.  

## 변경내역
- 기존에 구현한 RadioButton은 실제 HTML 태그상으로 아래와 같은 구조로 되어있었어요.
```HTML
<label>
  <input type="radio" /> // display: none;
  <div>
    // 시각적으로 표현되는 버튼
  </div>
</label>
```
하지만 이렇게 구현하는 버튼은 접근성을 위한 role도 다시 다 붙여 주어야 하고, 무엇보다 사용자가 이 버튼에 대한 `label` 태그를 사용하는 경우 useCase에 따라서는 `label` 태그가 중첩되는 모양새가 되는 문제가 있었어요. 사실 어떤 문제가 발생한다기 보다는, 스크린리더에 따라 이를 처리하는 방식이 다를 수 있다고 해요. 그래서 다음과 같이 변경했어요.
```HTML
<div> // outerSize zone
  <input type="radio" /> // outerSize zone의 크기 + 0 opacity
  <div aria-disabled>
    // 시각적으로 표현되는 버튼
  </div>
</div>
```
이렇게 하면 `label` 태그로 처리하던 `input-click` 동작을 `input` 태그가 직접 처리해서 접근성 측면에서 이점을 얻을 수 있어요.(제가 생각하기에 그런데 혹시 아니면 바로잡아주세요 😓)  

- inputRef 속성을 추가했어요.
  - 라디오 버튼에는 흔한 경우는 아니지만, input 태그를 `useRef`로 참조하는 것은 분명 가능해야 한다고 생각했어요. forwardRef로 돌려야 하나 했는데 그냥 간단하게 `RefObject<HTMLInputElement>`를 props로 받아 전달해주면 작동하길래 추가했어요.

## 고민하는 부분
> 요새 질문 글 쓰다보면 머릿속에 답이 생겨서 혼자 정리했어요. 그래도 혹시 의견이 있으시다면 알려주세요~
- customButton에 대한 생각
  - 자주 사용되는 디자인 시스템들의 커스텀 구현 방식을 보니, checked - idle - disabled 각각 컴포넌트를 받는 방법과 css 선택자를 사용하여 직접 구현할 것을 제안하는 방법, 또 data-attribute를 spec으로 제시하는 방법이 있었어요.
  - 처음에는 `checked`라는 className에 대해 대응하도록 구현할 것을 스펙으로 적어두고 구현하게 하려고 했는데, `input:checked ~ &` css 선택자를 사용할 것을 스펙에 적어두는 것과 크게 다르지는 않으면서 `RadioGroup`이 없으면 작동하지 않게 돼요.
- RadioGroup
  - `fieldset`에 대응하는 컴포넌트가 필요할까 고민하고 있었는데, 앞의 customButton도 그렇고 라디오버튼 자체가 group 없이 사용되는 경우가 없다고 생각해서(사실 저는 자주 그랬던 것 같은데 fieldset의 사용이 접근성 권고 사항이에요) 만드는게 좋을 것 같긴 해요.
  - 지금 생각하는 이 녀석의 역할은 customButton에 대응하기 위한 data-attribute(또는 className)을 넣어 주는 일 외에 떠오르는 건 없어요.
  - 그런데 이 경우에 fieldset을 스타일적으로 어떻게 구현해야 할지 고민돼요. 그냥 css props 받도록 해서 알아서 구현하도록 해도 될까요?
  - 그리고.. `legend`에 대응하는 컴포넌트도 필요할까요?
  - 사실 이제 다시 리뷰요청 열려고 했는데 RadioGroup 해야할 것 같아서 좀만 더 미룰게요 ㅠㅠ

## Refs.
https://nuli.navercorp.com/community/article/1132737  
https://mui.com/material-ui/react-radio-button  
https://chakra-ui.com/docs/components/radio

prayinforrain committed:
refactor: refObject props를 forwardRef로 변경

prayinforrain committed:
docs: ref 변경에 따른 스토리 수정

prayinforrain committed:
refactor: direction props 추가

prayinforrain committed:
refactor: self-closing tag로 사용 가능

prayinforrain committed:
feat: 키보드 조작 focus 표시

prayinforrain committed:
feat: 접근성 및 UI 가시성 개선

- 투명 input의 크기를 없애고 pointer-events: none 속성을 지움
- focus된 상태에서의 통일성 개선
- label에 cursor:pointer 적용

prayinforrain committed:
docs: 스토리 파일 수정

prayinforrain committed:
Merge branch 'main' into feat/radiobutton

prayinforrain committed:
refactor: label props를 줄 수 있도록 변경

prayinforrain commented:
> 위의 코멘트들은 이제 deprecated(?) 됐어요. 이것만 읽으셔도 될 것 같아요.

## 변경 내역
- refObject를 props로 받던 것을 forwardRef로 변경했어요.
  - props로 refObject를 받는 것이 클래스형 컴포넌트(물론 지금은 거의 안쓰지만)에서 문제를 일으킨다고 해요. 챗쌤이 한 말인데 검증 필요하긴 해요
- **다시 버튼 전체를 라벨로 감싸도록 변경되었어요.**
  - 처음에 언급한 라벨 중첩 문제는 라벨의 역할을 하는 요소까지 라디오 버튼이 포함하도록 변경하면서 해결(?) 했어요.
  - 회귀한 계기는 [이 댓글](https://github.com/c-h-w-h/cds/issues/42#issuecomment-1517408827)의 `pointer-events` 관련한 내용이에요.
- 라벨과 버튼의 위치 관계를 정의하는 direction props를 추가했어요.
  - 버튼을 기준으로 라벨의 위치를 정의해요. 기본은 `right`로, 버튼 오른 쪽에 라벨이 와요.
- 자식 요소를 받을 수 있어요. 즉, **self-closing 형태와 자식 요소를 갖는 형태 모두 사용이 가능해요.**
  - self-closing 형태에서는 label props가 주어지는 경우 label을, 그렇지 않은 경우 value 값을 라벨로 사용해요.
  - 자식 요소를 갖는 경우 해당 요소를 라벨로 사용해요.
- 키보드 focus가 사실 되고 있었는데 안되는 것으로 보였어요. 가시성을 조금 개선했어요.

## 생각
### 접근성 관련
오늘 하루 라디오 버튼에 관련된 접근성 문서들을 읽었는데.. 제가 제대로 이해한 게 맞다면 input 태그가 제대로 존재한다면 따로 aria 속성을 줄 필요는 없는 것으로 이해했어요. 왜냐면 `input type='radio'` 자체만으로 이미 충분히 시맨틱하기(?) 때문이에요. 다만 시각적으로 보이는 가짜 라디오 버튼은 `aria-hidden`을 주어 혼동이 없도록 했어요.

### RadioGroup 관련
같은 이유로 라디오그룹도 docs에만 언급하고 구현은 하지 않았어요.
- 가장 큰 이유는 굳이 추가로 컴포넌트를 구현해서 부여할만한 RadioGroup의 역할이 마땅히 없기 때문이였고
- 접근성 문서에서 `fieldset` - `legend` 태그를 사용하거나, aria 속성으로 이런 role이  부여된 태그를 사용할 것을 권장하고 있어서 처음에는 네임스페이스 맞추는 느낌으로 `styled.fieldset`을 반환하는 껍데기를 구현했어요.
  - 그런데 스토리를 작성하면서 사용해 보니 결국 개발자가 추가로 스타일을 넣는 것이 불가피한데 이걸 `<RadioGroup css={css어쩌고저쩌고} ... >` 하면서 props로 주입하는 게 더 가독성을 해치는 기분이 들었어요.
  - 따라서 개발자가 `emotion`으로 직접 구현하는 것이 더 좋은 DX라고 생각했어서 제외했어요. 만약 이 부분에 대한 디자인이 확정될 수 있다면 구현할 의향은 있어요.
- 레퍼런스를 찾아보니 다른 디자인 시스템들의 `RadioGroup` 컴포넌트들은 대부분 폼 데이터를 컨트롤하기 위한..? 폼 컨트롤 컴포넌트의 합성물로 보였어요.(아님 말고)

### 기타
제가 2주 전?에 라디오 버튼의 변종들에 대해서 이야기 했었는데.. 일단 지금 구조로도 라벨의 형태가 자유로워 졌기 때문에 이론상으로는(?) 구현이 가능해졌어요.

오랜만의 작업이라 뭔가 나사가 단단히 빠진 느낌이에요. 피드백 환영이에요 ㅠㅠ

se030 commented on src/components/RadioButton/index.tsx:
```diff
@@ -0,0 +1,158 @@
+    return (
+      <RadioButtonWrapper direction={direction}>
+        {(direction === 'left' || direction === 'top') && children}
+        <ActualButton
```
이건 ActualInput으로 네이밍하면 어떨까요?

se030 reviewed (approved):
고생하셨습니다! 👍🏻 children 인터페이스를 열어두신 이유가 자유로운 형태의 라디오 버튼을 고려하신거 맞나요?

VoiceOver 기준으로 self-closing 태그에서 라벨 값이 주어지지 않으면 보조 기술에 제공되는 정보가 없어요. aria-label을 달아주면 좋을 것 같아요. 뭔가 의도적으로 감추신 것 같지만 말씀드려요.
<img width="869" alt="image" src="https://user-images.githubusercontent.com/63814960/233780342-38554fe9-d6b8-411e-8243-15bfc0061edf.png">

outerSize라는 이름이 어떤 의도의 props인지 바로 이해되지 않았어요. clickableSize, interactiveSize 등으로 바꿔보면 어떨까요? 근데 이 부분의 커스텀을 열어두신 이유가 있나요? 

그리고 사소하지만 스타일드 컴포넌트들 네이밍(RadioWrapper, Button)에 태그명이 반영되면 더 좋을 것 같아요! 제가 짠 마크업이 아니다보니까 왔다갔다 하면서 매치하게 되더라고요.

se030 commented on src/components/RadioButton/index.tsx:
```diff
@@ -0,0 +1,158 @@
+          defaultChecked={checked}
+          disabled={disabled}
+          id={id}
+          ref={ref}
```
```tsx
        aria-label={label ?? value}
```

이 쪽에 이런식으로 넣으면 될 것 같아요. label, value 둘 다 optional인데 value은 required로 둬야하지 않을까요?

prayinforrain replied:
바닐라 HTML에서 optional이어서 그렇게 했는데 생각해보니 HTML은 필수가 없네요... 같이 반영할게요! 감사해요

prayinforrain commented:
> 고생하셨습니다! 👍🏻 children 인터페이스를 열어두신 이유가 자유로운 형태의 라디오 버튼을 고려하신거 맞나요?
> 
> VoiceOver 기준으로 self-closing 태그에서 라벨 값이 주어지지 않으면 보조 기술에 제공되는 정보가 없어요. aria-label을 달아주면 좋을 것 같아요. 뭔가 의도적으로 감추신 것 같지만 말씀드려요. <img alt="image" width="869" src="https://user-images.githubusercontent.com/63814960/233780342-38554fe9-d6b8-411e-8243-15bfc0061edf.png">
> 
> outerSize라는 이름이 어떤 의도의 props인지 바로 이해되지 않았어요. clickableSize, interactiveSize 등으로 바꿔보면 어떨까요? 근데 이 부분의 커스텀을 열어두신 이유가 있나요?
> 
> 그리고 사소하지만 스타일드 컴포넌트들 네이밍(RadioWrapper, Button)에 태그명이 반영되면 더 좋을 것 같아요! 제가 짠 마크업이 아니다보니까 왔다갔다 하면서 매치하게 되더라고요.

대부분 HTML 기본 요소로 접근성 속성들이 채워지다 보니 빠진 게 없나 계속 고민했는데 이런 문제가 있네요. 맥OS 보이스오버가 있는 걸 또 깜빡했어요. 한 번 쭉 확인해 볼게요 ㅋㅋ...
컴포넌트명은 다시 점검해 볼게요. 최초에 아무 생각 없이 짠 것이 그대로 유지된 것도 있고 사실 내부에서만 사용될 이름이라 대충 막 지었거든요.
그리고 children 인터페이스를 열어둔 이유는 말씀하신 것이 맞아요. 제가 이전에 조사했던 특수 케이스에 대응하기 위해 구현했는데 딱 '가능하기만' 하지 실제로 직관적일지는 잘 모르겠어요.

--추가
outerSize에 대한 의도를 답하는 걸 깜빡했는데 선민님한테 달았던 내용에 있어요!
>라디오버튼 자체가 크면 못생겼고, 작으면 클릭하기 불편하다는 이유였어요. hover했을 때의 변화가 인지하기 어려웠던 이유도 있고, 실제 클릭 가능한 영역을 더 넓게 가져가고 싶은 상황이 있을 거라고 생각했어요. 그리고 이것만 바꿀건데 커스텀 버튼으로 구현하려면 기본 버튼의 생김새까지 사용자가 새로 구현해야 해서.. ㅎㅎ

leesunmin1231 commented on src/components/RadioButton/index.tsx:
```diff
@@ -0,0 +1,158 @@
+  opacity: 0.5;
+`;
+
+RadioButton.displayName = 'RadioButton';
+
+export { RadioButton };
```
헐 이건 뭐에요? 이렇게 하면 default export 랑 똑같이 동작하는 건가요?? 처음봐요

prayinforrain replied:
forwardRef를 사용할 때 displayName을 지정해 주지 않으면 클래스형 컴포넌트에서 문제가 생긴다고 했나.. 암튼 챗쌤이 그랬던 것 같아요(?)

se030 replied:
헉 예리해요 default export로 바꾸고 배럴 파일에도 추가해주세요 ~

prayinforrain replied:
엥 앗 그런 얘기였군요 반영할게요 ㅋㅋㅋㅋㅋㅠㅠ 방금 일어나서 잘못 알아들었어요

leesunmin1231 commented on src/components/RadioButton/index.tsx:
```diff
@@ -0,0 +1,158 @@
+      checked = false,
+      color,
+      size = '16px',
+      outerSize = size,
```
음 outerSize까지 따로 커스텀 해서 사용할 필요가 있을까요??
라디오 버튼 size 정도면 custom 가능하게 하고 이 부분은 size에 따라 바꿔주는 정도가 디자인 통일성이 있어서 더 좋을 거 같아요.
어차피 customButton을 따로 받으니까, 기본 스타일은 정해두고 다른 스타일이 필요하면 사용자 필요에 따라 customButton을 따로 정의 해서 넣어주는게 좋은거 같아요.
개인적인 의견인데 이렇게 구현하신 이유가 있으신가요?

prayinforrain replied:
라디오버튼 자체가 크면 못생겼고, 작으면 클릭하기 불편하다는 이유였어요. hover했을 때의 변화가 인지하기 어려웠던 이유도 있고, 실제 클릭 가능한 영역을 더 넓게 가져가고 싶은 상황이 있을 거라고 생각했어요. 그리고 이것만 바꿀건데 커스텀 버튼으로 구현하려면 기본 버튼의 생김새까지 사용자가 새로 구현해야 해서.. ㅎㅎ

leesunmin1231 reviewed (approved):
우왕 고생하셨어요~!!
우재님의 깊은 고민의 흔적이 보여요 ㅎㅎ 수고하셨습니다!

iyu88 commented on src/components/RadioButton/RadioButton.stories.tsx:
```diff
@@ -0,0 +1,275 @@
+import { RadioButton } from '.';
+
+export default {
+  title: 'Design System/Components/RadioButton',
```
이거 이제 한 depth 줄이시면 됩니다!
'Design System' 을 제거해주세요~

prayinforrain committed:
refactor: label에 대한 접근성 피드백 반영

iyu88 commented on src/components/RadioButton/RadioButton.stories.tsx:
```diff
@@ -0,0 +1,275 @@
+  },
+  argTypes: {
+    size: {
+      name: 'size',
```
argTypes에서 name은 추론이 잘 되어서 모두 제거해주면 좋을 것 같아요!

iyu88 commented on src/components/RadioButton/RadioButton.stories.tsx:
```diff
@@ -0,0 +1,275 @@
+      name: 'size',
+      description: '버튼의 크기를 지정합니다.',
+      table: {
+        type: { summary: 'string' },
```
원래는 

control: {
        type: "number",
},

이렇게 컨트롤 부분을 제안드리려고 했는데 type이 string 이네요

CSS 속성으로 px이든 rem이든 받게 하고 싶으신 의도라고 파악했는데 맞을까요?

만약 그렇다면

`summary: 'CSSProperties['width']'`

처럼 설정하고 지금처럼 control 을 막는 것도 방법이 될 수 있겠다고 생각했습니다

type에 일반 string이 들어가 있는게 어색해보여서요!

prayinforrain replied:
>CSS 속성으로 px이든 rem이든 받게 하고 싶으신 의도라고 파악했는데 맞을까요?

이게 맞아요. 저 옵션으로 control의 타입 추론이 이루어지나 해서 원시형인 string을 넣었는데 그냥 읽는 사람에게 보여주기 위한 값인 것 같으니 CSSProperties로 수정할게요!  
지금 control을 막아두지는 않았는데 혹시 제가 보고 있는 것과 다르게 보이시나요?  

![image](https://user-images.githubusercontent.com/46566891/233851530-e5681d09-56b5-4ff4-83ac-afba97d8a8c0.png)  
(*size 외에 일부 피드백이 반영된 상태임)

iyu88 commented on src/components/RadioButton/RadioButton.stories.tsx:
```diff
@@ -0,0 +1,275 @@
+      name: 'outerSize',
+      description: '버튼의 클릭 영역 크기를 지정합니다.',
+      table: {
+        type: { summary: 'string' },
```
여기도 유지한다면 

`summary: 'CSSProperties['width']'`

로 바꿀 수 있을 것 같습니다

iyu88 commented on src/components/RadioButton/RadioButton.stories.tsx:
```diff
@@ -0,0 +1,275 @@
+      description: '기본 버튼의 색상을 지정합니다.',
+      control: {
+        type: 'color',
+      },
```
순서만 table 밑으로 내려주세요~

prayinforrain committed:
style: 태그명 직관성 개선

prayinforrain committed:
refactor: export default로 변경 및 배럴 파일에 추가

prayinforrain committed:
refactor: 직관적이지 못한 prop명 수정

iyu88 commented on src/components/RadioButton/RadioButton.stories.tsx:
```diff
@@ -0,0 +1,276 @@
+      },
+    },
+    name: {
+      table: {
+        disable: true,
+      },
```
나머지 props들의 description 이나 type 을 명시하지 않은 이유가 있을까요?
스토리에서 확인할 때 control 을 제공하지 않더라도 설명은 있으면 어떤 props가 필요한지 
파악하기 좋을 것 같다고 생각했어요~

prayinforrain replied:
개발 과정에서 불필요한 props들이(예를 들면 바닐라 HTML의 input props라던가..) 마구 표시되어 지워두었는데 라디오 버튼에 대한 props 일부는 살려서 description 추가할게요. 너무 작업 텀이 길어지다 보니까 자꾸 디테일을 놓치네요😅  
추가하려는 props는 아래와 같아요.

- label (얘는 원래 추가해야 했던 오리지널 prop이에요)
- disabled
- customButton(컨트롤 막고 설명만)

~아래로는 고민중인 것들이에요. 근데 안넣기로 혼자 정했어요. 위에 것들 추가하니 너무 길어져서..
- checked
- value
- name

iyu88 commented on src/components/RadioButton/RadioButton.stories.tsx:
```diff
@@ -0,0 +1,276 @@
+      },
+    },
+  },
+  args: {
+    size: '16px',
+    disabled: false,
+    color: theme.color.primary100,
+    clickableSize: '30px',
+    direction: 'right',
```
이렇게 지정하는 방법이 있군요
대박이에요 👍

prayinforrain replied:
앗 맞다 이거 이렇게 기본값 넣어줘야 control에서 멀쩡(?)하게 조작 가능하더라구요. 이게 없는 string props는 Text Input 대신 `Set String` 버튼이 나와가지구.. 이거 다음 회의 때 한번 얘기해 봐야겠다 생각했었어요

iyu88 commented on src/components/RadioButton/RadioButton.stories.tsx:
```diff
@@ -0,0 +1,276 @@
+        defaultValue: { summary: '16px' },
+      },
+    },
+    clickableSize: {
```
이름 변경 좋습니다 👍

iyu88 commented on src/components/RadioButton/RadioButton.stories.tsx:
```diff
@@ -0,0 +1,276 @@
+        disable: true,
+      },
+    },
+  },
```
뭔가 이쯤에 decorator로 중앙 정렬되면 깔끔할 것 같은데 어떻게 생각하시나요 ㅎㅎ

iyu88 commented on src/components/RadioButton/RadioButton.stories.tsx:
```diff
@@ -0,0 +1,276 @@
+Default.args = {};
+
+export const CustomColor = Template.bind({});
+CustomColor.args = {
```
설명이 없는 스토리에도 간단하게 

'어떤 props를 어떻게 조정한 것인지' 에 대한 storyDescription 이 있으면 

다른 Docs 와 통일성이 있을 것 같아요!

iyu88 commented on src/components/RadioButton/index.tsx:
```diff
@@ -0,0 +1,159 @@
+  children?: ReactNode;
+}
+
+const RadioButton = forwardRef<HTMLInputElement, RadioButtonProps>(
```
이렇게 하면 사용할 때 ref로 

RadioButton을 참조할 수 있게 되는 것이겠지요?? 😮

prayinforrain replied:
맞습니당~ ref props로 넘어오는 refObject는 ActualInput(`input[type='radio']`)로 내려줍니당

iyu88 commented on src/components/RadioButton/index.tsx:
```diff
@@ -0,0 +1,159 @@
+  justify-content: center;
+  position: relative;
+
+  & > input[type='radio']:not(:disabled) ~ * {
+    cursor: pointer;
```
'radio가 disabled가 아니면 모든 것에 대해서 cursor: pointer 한다'

가 맞나요? 🙄

prayinforrain replied:
맞아요! disabled되어 있는데 pointer 커서가 되는 게 거슬려서 수정했어요

se030 replied:
disabled 상태일 때 cursor: not-allowed 를 써보면 어때요?

prayinforrain replied:
not-allowed 커서가 익숙하지 않아서 생각도 못했는데 알려주셔서 방금 넣었어요~

iyu88 commented on src/components/RadioButton/index.tsx:
```diff
@@ -0,0 +1,159 @@
+    if (!color) color = themeColor.primary100;
+
+    return (
+      <WrapperDiv direction={direction}>
```
WrapperDiv가 사실은 label이 맞겠죠?
이름에 div가 들어가 있어서 그런지 헷갈려요..!
Div가 이름에 포함되지 않아도 될 것 같은데 이유가 있을까요?

prayinforrain replied:
현빈님 말이 맞아요.. 사실 잘못 넣은 거에요.. 수정할게요 😥

iyu88 reviewed (approved):
수고 많으셨습니다!
하나 궁금한 점은 아래 사진에서 버튼과 텍스트 사이에 간격은 조절할 수 있을까요?
커스텀으로 넣는다기보다 임의로 살짝 간격이 있으면 좋을 것 같다는 의견입니다~
고생하셨어요~~

<img width="242" alt="스크린샷 2023-04-23 23 16 24" src="https://user-images.githubusercontent.com/31645195/233845068-16a695ba-a4ab-4762-8d15-663c6aff13b1.png">

prayinforrain committed:
docs: 스토리 설명 보강

prayinforrain committed:
fix: storybook showCode에 No Display Name이 출력됨

prayinforrain committed:
fix: 잘못 지정된 컴포넌트명 수정

prayinforrain committed:
docs: 스토리 관련 피드백 반영

- 카테고리 depth 변경
- size props 타입 명시
- argTypes 옵션 순서 수정
- 설명이 필요한 props 추가

prayinforrain commented:
>하나 궁금한 점은 아래 사진에서 버튼과 텍스트 사이에 간격은 조절할 수 있을까요?
커스텀으로 넣는다기보다 임의로 살짝 간격이 있으면 좋을 것 같다는 의견입니다~

현빈님 피드백을 듣고 `gap` props를 추가했습니다!

prayinforrain committed:
feat: gap props 추가

prayinforrain committed:
docs: disabled argTypes 순서 정리

prayinforrain committed:
feat: disabled 버튼에 커서 스타일 추가

se030 commented:
> > 하나 궁금한 점은 아래 사진에서 버튼과 텍스트 사이에 간격은 조절할 수 있을까요?
> > 커스텀으로 넣는다기보다 임의로 살짝 간격이 있으면 좋을 것 같다는 의견입니다~
> 
> 현빈님 피드백을 듣고 `gap` props를 추가했습니다!

이 커스텀을 열어둘 필요는 없어보여요 ...!! 오히려 일관적인 형태 제공이라는 목적을 해치는 것 같아서 clickable 값을 고려한 고정값으로 두면 어떨까 싶어요.

prayinforrain commented:
헉 그렇다면 내일 오전까지 고민해보고 뺄게요! 추가 수정이 또 생길지도 모르니..
뭔가 크로마틱 빌드수 아끼려고 push 늦추게 돼요

prayinforrain committed:
refactor: gap props 제거

prayinforrain committed:
Merge branch 'main' into feat/radiobutton
